src/Control/Carrier/TaskPool.hs

```diff
@@ -101,16 +114,17 @@ data Progress = Progress
 newtype TaskPoolC m a = TaskPoolC { runTaskPoolC :: ReaderC (TVar [m ()]) m a }
   deriving (Functor, Applicative, Monad, MonadIO, MonadFail)
 
-instance forall sig m. (Algebra sig m, MonadIO m) => Algebra (TaskPool :+: sig) (TaskPoolC m) where
+instance (Has (Lift IO) sig m, Algebra sig m) => Algebra (TaskPool :+: sig) (TaskPoolC m) where
   alg hdl sig ctx = TaskPoolC $ case sig of
     L (ForkTask m) -> do
       var <- ask @(TVar [m ()])
-      liftIO $ atomically $ modifyTVar' var (runReader var (runTaskPoolC (void (hdl (m <$ ctx)))):)
+      sendIO $ atomically $ modifyTVar' var (runReader var (runTaskPoolC (void (hdl (m <$ ctx)))):)
       pure ctx
     R other -> alg (runTaskPoolC . hdl) (R other) ctx
 
 data State m = State
   { stQueued    :: TVar [m ()]
   , stRunning   :: TVar Int
   , stCompleted :: TVar Int
+  , stError     :: TMVar SomeException
```

Comment:
This is the trick: we have a mutable variable that's either empty or it contains a single exception. tasks insert an exception when one is thrown